packages/flutter/lib/src/material/contained_button.dart

```diff
@@ -0,0 +1,425 @@
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:math' as math;
+import 'dart:ui' show lerpDouble;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+import 'button_style.dart';
+import 'button_style_button.dart';
+import 'color_scheme.dart';
+import 'constants.dart';
+import 'contained_button_theme.dart';
+import 'material_state.dart';
+import 'theme.dart';
+import 'theme_data.dart';
+
+/// A Material Design "contained button".
+///
+/// Use contained buttons to add dimension to otherwise mostly flat
+/// layouts, e.g.  in long busy lists of content, or in wide
+/// spaces. Avoid using contained buttons on already-contained content
+/// such as dialogs or cards.
+///
+/// A contained button is a label [child] displayed on a [Material]
+/// widget whose [Material.elevation] increases when the button is
+/// pressed. The label's [Text] and [Icon] widgets are displayed in
+/// [style]'s [ButtonStyle.onForegroundColor] and the button's filled
+/// background is the [ButtonStyle.backgroundColor].
+///
+/// The contained button's default style is defined by
+/// [defaultStyleOf].  The style of this contained button can be
+/// overridden with its [style] parameter. The style of all contained
+/// buttons in a subtree can be overridden with the
+/// [ContainedButtonTheme], and the style of all of the contained
+/// buttons in an app can be overridden with the [Theme]'s
+/// [ThemeData.containedButtonTheme] property.
+///
+/// The static [styleFrom] method is a convenient way to create a
+/// contained button [ButtonStyle] from simple values.
+///
+/// If [onPressed] and [onLongPress] callbacks are null, then the
+/// button will be disabled.
+///
+/// See also:
+///
+///  * [TextButton], a simple flat button without a shadow.
+///  * [OutlinedButton], a [TextButton] with a border outline.
+///  * <https://material.io/design/components/buttons.html>
+class ContainedButton extends ButtonStyleButton {
+  /// Create a ContainedButton.
+  ///
+  /// The [autofocus] and [clipBehavior] arguments must not be null.
+  const ContainedButton({
+    Key key,
+    @required VoidCallback onPressed,
+    VoidCallback onLongPress,
+    ButtonStyle style,
+    FocusNode focusNode,
+    bool autofocus = false,
+    Clip clipBehavior = Clip.none,
+    @required Widget child,
+  }) : super(
+    key: key,
+    onPressed: onPressed,
+    onLongPress: onLongPress,
+    style: style,
+    focusNode: focusNode,
+    autofocus: autofocus,
+    clipBehavior: clipBehavior,
+    child: child,
+  );
+
+  /// Create a contained button from a pair of widgets that serve as the button's
+  /// [icon] and [label].
+  ///
+  /// The icon and label are arranged in a row and padded by 12 logical pixels
+  /// at the start, and 16 at the end, with an 8 pixel gap in between.
+  ///
+  /// The [icon] and [label] arguments must not be null.
+  factory ContainedButton.icon({
```

Comment:
These two constructors end up being very similar. Do you think it would be reasonable to consider combining `ContainedButton` and `ContainedButton.icon` into one constructor? The constructor could have an optional (nullable) `icon` param, that, when non null, will place itself alongside `child`?

Reply:
FWIW, the Foo and Foo.icon constructors are how this was factored before. We're not trying for API compatibility here, so there's no need to include the factory again. On the other hand, Foo and Foo.icon have different defaultStyleOf overrides because their default padding is different and because the padding's dependency on textScaleFactor differs. This and the difference in how the child is handled could certainly be blended together into one build method, but separating the two cases with a class (albeit a private one) simplifies that a little.